### ISSUE TITLE
New plugin manager does not load custom javascript for built-in pages

### ISSUE THREAD

**Issue by Zen4All** [zencart/zencart#3563]
I am testing the new plugin manager, and I know it is not 100% ready yet, but it is a new toy and I want to play with it 😄 .

In testing I am converting a plugin to use the new structure. Most files load ok, but the files I have put in ```zc_plugins/ProfitMarginCalculator/admin/includes/javascript``` are not loaded. It seems like the javascript_loader does not know it is there.
Do we simply need to add a path somewhere, or is this beyond the capabilities of the plugin manager.

**Comment by zcwilt** [comment]
Funnily enough @drbyte and I were discussing asset management and plugins last night.

At the moment there is no autoloading of assets related to plugins but that will come in v1.5.8. 
However for now I'll create a helper function to get the path to the plugin, which can be used to load css/js within the plugin.

**Comment by zcwilt** [comment]
btw I assume the path you posted above is a mistake and should include the version directory.

**Comment by Zen4All** [comment]
> Funnily enough @drbyte and I were discussing asset management and plugins last night.
> 
> At the moment there is no autoloading of assets related to plugins but that will come in v1.5.8.
> However for now I'll create a helper function to get the path to the plugin, which can be used to load css/js within the plugin.

thx

**Comment by Zen4All** [comment]
> btw I assume the path you posted above is a mistake and should include the version directory.

yes

**Comment by zcwilt** [comment]
k, so haven't pushed the PR yet. Still testing
but the way it would work would be something like 

`$pluginBaseDir = Filesystem::getInstance()->getPluginRelativeDirectory('DisplayLogs', $installedPlugins);`

and note here you would need to have a `use Zencart\FileSystem\FileSystem;` at the top of the file 

you can then do something like

```
<link rel="stylesheet" href="<?php echo $pluginBaseDir; ?>admin/includes/assets/myscript.css">
<script src="<?php echo $pluginBaseDir; ?>admin/includes/assets/myscript.js"></script>

```

This will probably just be a temporary solution, as I mentioned above that asset management probably needs a lot more infrastructure.

**Comment by zcwilt** [comment]
In fact I would like us (the core team) and the community come up with some guideline standards around plugin structure. 
e.g. standardizing asset directories etc.

**Comment by Zen4All** [comment]
:+1:  Will het back to this after dinner

**Comment by Zen4All** [comment]
There are two obvious directories:
- admin/includes/javascript
- admin/includes/css

For some mods I use
- admin/includes/modals

**Comment by zcwilt** [comment]
Have simplified this so now you just need to do

`$pluginBaseDir = Filesystem::getInstance()->getPluginRelativeDirectory('DisplayLogs');`

**Comment by Zen4All** [comment]
This will work in a new admin page that is in the plugin directory, but not in existing admin pages. (Without editing them)

I wanted to add a JavaScript to product.php, but for now I can't see a way to do that without editing core files, or adding files outside of the plugin manager structure.

I hope you will add a possibility to auto-load JavaScript and CSS files in a future version for use in core files.

**Comment by zcwilt** [comment]
Definitely the idea would be to allow this. 
As mentioned earlier, it needs to manage assets for css/js and images 
I am working on this.

**Comment by zcwilt** [comment]
I would of preferred to do this in v158 using some new resource loader classes, as there are also issues related to loading images local to a plugin as well. zen_image needs retiring and rewriting.

For now though will try and do a very quick update for admin to allow this.

**Comment by zcwilt** [comment]
Have pushed #3581 

This allows for some auto loading of css/js

e.g. 
if 
zc_plugins/DisplayLogs/v3.0.0/admin/includes/css/display_logs.css
or 
zc_plugins/DisplayLogs/v3.0.0/admin/includes/javascript/display_logs.js

exists, they will be loaded

It doesn't allow for `directory` based loading e.g. 

zc_plugins/DisplayLogs/v3.0.0/admin/includes/javascript/display_logs_1.js
zc_plugins/DisplayLogs/v3.0.0/admin/includes/javascript/display_logs_2.js

etc.

**Comment by Zen4All** [comment]
@zcwilt 
Almost there 😄 
I changed your code (locally) in javavascript_loader to:
```
foreach ($installedPlugins as $plugin) {
    $relativeDir = $fs->getPluginRelativeDirectory($plugin['unique_key']);
    $absoluteDir = $fs->getPluginAbsoluteDirectory($plugin['unique_key']);
    if (file_exists($absoluteDir . 'admin/includes/javascript/' . basename($PHP_SELF, '.php') . '.php')) {
        echo "\n";
        require $absoluteDir . 'admin/includes/javascript/' . basename($PHP_SELF, '.php') . '.php';
    }
    if (file_exists($absoluteDir . 'admin/includes/javascript/' . basename($PHP_SELF, '.php') . '.js')) {
        echo "\n";
?>
        <script src="<?php echo $relativeDir ?>admin/includes/javascript/<?php echo basename($PHP_SELF, '.php') . '.js'; ?>"></script>
<?php 
    }
    $directory_array = $template->get_template_part($absoluteDir . 'admin/includes/javascript/', '/^' . basename($PHP_SELF, '.php') . '_/', '.js');
foreach ($directory_array as $key => $value) {
    echo "\n";
?>
<script src="<?php echo $relativeDir; ?>admin/includes/javascript/<?php echo $value; ?>"></script>
<?php
}
$directory_array = $template->get_template_part($absoluteDir . 'admin/includes/javascript/', '/^' . basename($PHP_SELF, '.php') . '_/', '.php');
foreach ($directory_array as $key => $value) {
    echo "\n";
    require $absoluteDir . 'admin/includes/javascript/' . $value;
}
}
```
and admin_html_head to:
```
foreach ($installedPlugins as $plugin) {
    $relativeDir = $fs->getPluginRelativeDirectory($plugin['unique_key']);
    $absoluteDir = $fs->getPluginAbsoluteDirectory($plugin['unique_key']);
    if (file_exists($absoluteDir . 'admin/includes/css/' . basename($PHP_SELF, '.php') . '.css')) {
?>
        <link rel="stylesheet" href="<?php echo $relativeDir . 'admin/includes/css/' . basename($PHP_SELF, '.php') . '.css'; ?>">
<?php
    }
    $directory_array = $template->get_template_part($absoluteDir . 'admin/includes/css/', '/^' . basename($PHP_SELF, '.php') . '_/', '.css');
foreach ($directory_array as $key => $value) {
?>
    <link rel="stylesheet" href="<?php echo $relativeDir . 'admin/includes/css/<?php echo $value; ?>">
<?php
}
}
```
Now multiple custom css and javascripts will load.

**Comment by drbyte** [comment]
@Zen4All  a tip:

Using the backticks to wrap code samples is good. You can enhance it even further by specifying a language and it'll do syntax-highlighting for that language:

eg:
<img width="363" alt="Screen Shot 2020-06-09 at 3 45 04 PM" src="https://user-images.githubusercontent.com/404472/84192612-3d4d0580-aa68-11ea-9902-fdcf5c4ce3a8.png">
renders as:
```php
if (IS_ADMIN_FLAG === true) echo "I'm in the Admin";
```

Even fancier, if you use `diff` for the language it'll let you show your code as a diff (you'll have to manually add the `-` and `+` to the beginning of the lines, and show both old/new lines of course):

<img width="393" alt="Screen Shot 2020-06-09 at 3 46 20 PM" src="https://user-images.githubusercontent.com/404472/84192715-666d9600-aa68-11ea-8c6f-a87222e0bab1.png">
renders as:

```diff
-if (IS_ADMIN_FLAG === true) echo "I'm in the Admin";
+if (IS_ADMIN_FLAG !== true) echo "I'm not in the admin";
```

**Comment by zcwilt** [comment]
Should note as well there is still a piece missing here and that is to allow some global js/css in plugins
i.e. js/css that loads on every page.

**Comment by drbyte** [comment]
Closed in #3582
